src/lsm/groove.zig

```diff
@@ -580,34 +581,89 @@ pub fn GrooveType(
             assert(groove.prefetch_ids.count() == 0);
         }
 
+        /// May be called on a Groove for which has_id is true.
+        ///
+        /// This function makes a best-effort attempt to ascertain the `absence` of the passed key
+        /// in the IdTree, without performing I/O. If the absence of the key cannot be ascertained
+        /// without I/O, it is enqueued for prefetching.
+        pub fn prefetch_enqueue_negative_lookup_optimized(groove: *Groove, key: PrimaryKey) void {
```

Comment:
My rationale behind keeping them separate was:

1. Doing `groove.ids.manifest_search(key)` pre-emptively is favorable for negative lookups, but may not be for positive lookups - The purpose of negative lookups is to eagerly try to reject the existence of a key in the tree, which is why the current version attempts to bail out if a key doesn't exist in the manifest, or enqueue the key for prefetching. However, for positive lookups, the common case would be to find an existing key from the Grid cache (if not found in the grid cache), which is why I put the calls to `groove.ids.manifest_search(key)/groove.objects.manifest_search(key)` in the `PrefetchWorker`, so they can benefit from the parallelism of the PrefetchWorker to synchronously fetch blocks from the Grid cache.
2. `groove.ids.key_may_exist_in_tree(key)` benefits negative lookups, but not positive lookups - In the common case, positive lookups would involve fetching keys that do exist in the tree, so the check seemed wasteful for positive lookups. However, for negative lookups, the tree shape can help quickly ascertain the absence of a key in the tree.

It would be interesting to back this up with numbers, though! I shall run the microbenchmarks after adding `groove.ids.key_may_exist_in_tree(key)` to `prefetch_enqueue` and invoking it for negative lookups as well as positive lookups. Will post the numbers here!

Reply:
Ok, thanks. What are the microbenchmarks you are referring to?

Reply:
@sentientwaffle: I have some local changes wherein I've placed a timer to measure the performance of only the prefetch flow! I'm using the workload in `benchmark.zig` and I ensure that all reads happen from the Grid cache (by placing a panic in the grid's read path.

I used the aforementioned strategy to get these numbers: https://github.com/tigerbeetle/tigerbeetle/pull/944#issuecomment-1631196954

Reply:
Resolution: Based on the numbers in https://github.com/tigerbeetle/tigerbeetle/pull/944#issuecomment-1632973930, we have unified positive & negative prefetch flows, i.e. `prefetch_enqueue_negative_lookup_optimized` has been merged into `prefetch_enqueue.`